Tests/SwiftProtobufTests/Test_AsyncMessageSequence.swift

```diff
@@ -0,0 +1,241 @@
+// Tests/SwiftProtobufTests/Test_AsyncMessageSequence.swift - 
+//
+// Copyright (c) 2023 Apple Inc. and the project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See LICENSE.txt for license information:
+// https://github.com/apple/swift-protobuf/blob/main/LICENSE.txt
+//
+// -----------------------------------------------------------------------------
+///
+/// Tests the decoding of binary-delimited message streams, ensuring various invalid stream scenarios are
+/// handled gracefully.
+///
+// -----------------------------------------------------------------------------
+
+import Foundation
+import XCTest
+import SwiftProtobuf
+
+class Test_AsyncMessageSequence: XCTestCase {
+  
+  // Decode a valid binary delimited stream
+  func testValidSequence() async throws {
+    let expected: [Int32] = Array(1...5)
+    var messages = [SwiftProtoTesting_TestAllTypes]()
+    for messageNumber in expected {
+      let message = SwiftProtoTesting_TestAllTypes.with {
+        $0.optionalInt32 = messageNumber
+      }
+      messages.append(message)
+    }
+    let url = temporaryFileURL()
+    try writeMessagesToFile(url, messages: messages)
```

Comment:
Once your type becomes generic over the base async sequence you don't have to write to files at all in tests but can just construct the byte stream in process via something like `AsyncStream` which should work way better.

Reply:
Agreed in principle, but the current binary delimited serialization is limited to writing to `OutputStream`.  AFAIK it's necessary to do something horrible like:
```
    // See https://bugs.swift.org/browse/SR-5404
    let nsData = memoryOutputStream.property(forKey: .dataWrittenToMemoryStreamKey) as! NSData
    return Data(referencing: nsData)
```
which is worse IMO. Something to revisit maybe?

Reply:
Ah interesting I thought we can write binary delimited data into `Data` as well.